### PR TITLE
Adding arbitraries

### PR DESCRIPTION
**Pull request by nhibberd**
Not sure where else to put an orphan like this?

### CODE REVIEW

**Comment by markhibberd** [comment]
This is a good spot, but can you just add a prop that just does something relatively dumb to verify behaviour, maybe something like: `prop((c: Codec) => Codec(c.toString) ==== c)` (if that passes?). This is more to put a stake in the ground to indicate its use then verifying anything, but it is still useful.

**Comment by markhibberd** [comment]
:confetti_ball: After the spec.

**Comment by charleso** [comment]
Seems like a good place to me. :+1:

**Comment by nhibberd** [comment]
@markhibberd @charleso

File: src/main/scala/com/ambiata/disorder/ListS.scala [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+/**
+  This is a list of 'String' where there is no new
+  line characters.
+  */
+case class ListS(value: List[String])
+
+object ListS {
+  implicit def ListSArbitrary: Arbitrary[ListS] =
+    Arbitrary(for {
+      a <- arbitrary[List[S]]
```
Comment (charleso) [inline src/main/scala/com/ambiata/disorder/ListS.scala]: Would probably just `arbitrary[List[S]].map(...)` here /supernit

File: src/main/scala/com/ambiata/disorder/GenPlus.scala [hunk 2]
```diff
@@ -18,4 +18,26 @@ object GenPlus {
...
+  def checkGen[A](n: Int, gen: Gen[A], label: String, check: A => Boolean): Gen[A] = {
+    if (n > 100) Gen.fail.label(s"Couldn't generate a $label")
+    else for {
+      a <- gen
+      z <- Gen.size
+      r <- if (check(a)) Gen.const(a) else Gen.resize(z + 1, checkGen[A](n + 1, gen, label, check))
+    } yield r
+  }
+
+  def checkGenTuple[A, B: Arbitrary, C: Arbitrary](make: (B, C) => A, label: String, check: A => Boolean): Gen[A] = for {
```
Comment (charleso) [inline src/main/scala/com/ambiata/disorder/GenPlus.scala]: Hmm. Not sure this combinator is buying you much TBH - just call `checkGen` when a gen of tuple or whatever you need. /justmyopinion

File: src/main/scala/com/ambiata/disorder/EncodingString.scala [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+case class EncodingString(value: String, codec: Codec)
+
+object EncodingString {
+  implicit def EncodingStringArbitrary: Arbitrary[EncodingString] =
+    Arbitrary(checkGenTuple[EncodingString, S, Codec](
+        (s, c) => EncodingString(s.value, c)
+      , "EncodingString"
+      , es => validForCodec(es.value, es.codec)))
+
+  def validForCodec(s: String, c: Codec): Boolean =
```
Comment (charleso) [inline src/main/scala/com/ambiata/disorder/EncodingString.scala]: As mentioned IRL, the alternative is we define the supported/known character ranges for UTF8 and ISO8859. Not sure if that's better or worse though?

File: src/test/scala/com/ambiata/disorder/ListSSpec.scala [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+import org.specs2.{ScalaCheck, Specification}
+
+class ListSSpec extends Specification with ScalaCheck { def is = s2"""
+
+  ListS properties are met:
+
+    ${ prop((s: ListS) => s.value.filterNot(_.contains("\\s")).length ==== s.value.length) }
+
+    ${ prop((s: ListS) => s.value.filterNot(_.contains("\\n")).length ==== s.value.length) }
```
Comment (charleso) [inline src/test/scala/com/ambiata/disorder/ListSSpec.scala]: I'm a little confused - this is the (literal) "\n" string? Is this what you mean to check?

**Comment by charleso** [comment]
:hamburger:

File: src/main/scala/com/ambiata/disorder/GenPlus.scala (hunk 2 above)
Comment (nhibberd) [inline src/main/scala/com/ambiata/disorder/GenPlus.scala]: Saves a bunch of duplicate client code in DistinctPair/EncodingString for example

File: src/test/scala/com/ambiata/disorder/ListSSpec.scala (hunk 4 above)
Comment (nhibberd) [inline src/test/scala/com/ambiata/disorder/ListSSpec.scala]: The string contains the "\n" yeah?

File: src/main/scala/com/ambiata/disorder/EncodingString.scala (hunk 3 above)
Comment (markhibberd) [inline src/main/scala/com/ambiata/disorder/EncodingString.scala]: https://github.com/webcrank/webcrank-test.scala/blob/master/src/main/scala/webcrank/test/UnicodeArbitrary.scala is how to do it for UTF-8. Not sure how you would go replicating. The only reason to do one over the other is performance IMO, the currently implemented solution avoids the pitfalls mentioned in the link.

Comment (charleso) [inline src/main/scala/com/ambiata/disorder/EncodingString.scala]: > The only reason to do one over the other is performance IMO, the currently implemented solution avoids the pitfalls mentioned in the link.

Yeah was just thinking out loud really. If performance isn't an issue with what we have is certainly much easier.

File: src/main/scala/com/ambiata/disorder/ListS.scala [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+    } yield ListS(a.map(_.value.replaceAll("\\s", ""))))
```
Comment (markhibberd) [inline src/main/scala/com/ambiata/disorder/ListS.scala]: `replace("\n")`

File: src/test/scala/com/ambiata/disorder/ListSSpec.scala [hunk 6]
```diff
@@ -0,0 +1,16 @@
+package com.ambiata.disorder
+
+import org.specs2.{ScalaCheck, Specification}
+
+class ListSSpec extends Specification with ScalaCheck { def is = s2"""
+
+  ListS properties are met:
+
+    ${ prop((s: ListS) => s.value.filterNot(_.contains("\\s")).length ==== s.value.length) }
```
Comment (markhibberd) [inline src/test/scala/com/ambiata/disorder/ListSSpec.scala]: Either contains -> matches or contains("\n").

File: src/main/scala/com/ambiata/disorder/ListS.scala [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+  This is a list of 'String' where there is no new
+  line characters.
+  */
+case class ListS(value: List[String])
+
+object ListS {
+  implicit def ListSArbitrary: Arbitrary[ListS] =
+    Arbitrary(for {
+      a <- arbitrary[List[S]]
+      c <- arbitrary[Char]
```
Comment (markhibberd) [inline src/main/scala/com/ambiata/disorder/ListS.scala]: Check it is not a new line.

**Pull request merged**